feat: Add theme support to Stac framework with cloud loading

Potatomonsta opened the pull request:
## Description

This PR adds comprehensive theme support to the Stac framework, enabling themes to be loaded from Stac Cloud similar to how screens are currently loaded. The implementation includes async theme loading, intelligent caching, and proper context handling.

### Key Changes

**New Features:**
- Added `StacArtifactType` enum to support both screens and themes as fetchable artifacts
- Added `StacCloudTheme` annotation for marking theme builder methods
- Added `StacAppTheme` helper class with methods to load themes from:
  - Stac Cloud (`fromCloud`)
  - Network requests (`fromNetwork`)
  - Raw JSON (`fromJson`)
- Extended `StacApp` to support `FutureOr<StacTheme?>` for both `theme` and `darkTheme` parameters, enabling async theme loading
- Added loading state handling with `_ThemeFutureLoading` widget when themes are being fetched

**Refactoring:**
- Refactored `StacCacheService` to be artifact-type agnostic:
  - Renamed methods from screen-specific to artifact-specific (`getCachedArtifact`, `saveArtifact`, etc.)
  - Added artifact type parameter to all cache operations
  - Supports both `'screen'` and `'theme'` artifact types
- Refactored `StacCloud` to support fetching both screens and themes:
  - Unified fetching logic in `_fetchArtifact` method
  - Added `fetchTheme` method as public API
  - Updated background fetch tracking to support multiple artifact types
  - Added theme-specific cache management methods (`clearThemeCache`, `clearAllThemeCache`)

## Type of Change

<!--- Put an `x` in all the boxes that apply: -->

- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Code refactor
- [ ] Build configuration change
- [ ] Documentation
- [ ] Chore

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Multi-source theming (DSL/cloud/network/JSON) via a StacAppTheme wrapper
  * Theme annotation for runtime registration
  * Artifact type enum to classify cached items
  * Cloud APIs to fetch/clear themes

* **Refactor**
  * App now resolves themes asynchronously and shows a loading indicator
  * Cache model renamed to artifact-aware StacCache with per-type cache APIs

* **Documentation**
  * Expanded theming guide and added theming page to docs navigation

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Potatomonsta committed:
feat: Add theme support to Stac framework with cloud loading

Add comprehensive theme support to the Stac framework, enabling themes to be loaded from Stac Cloud similar to screens. This includes:

- Add StacArtifactType enum to support both screens and themes
- Add StacCloudTheme annotation for marking theme builder methods
- Add StacAppTheme helper class for loading themes from cloud, network, or JSON
- Extend StacApp to support FutureOr<StacTheme?> for async theme loading
- Refactor StacCacheService to be artifact-type agnostic (supports screens and themes)
- Refactor StacCloud to support fetching both screens and themes with unified caching

Themes can now be loaded asynchronously from Stac Cloud using the same caching strategies as screens (optimistic, cache-first, network-first, etc.), with proper loading states handled in StacApp.

divyanshub024 commented on packages/stac_core/lib/annotations/stac_theme_ref.dart:
````diff
@@ -0,0 +1,19 @@
+/// This annotation is used to identify Stac theme builders so the framework can
+/// register them and apply the correct theme at runtime.
+///
+/// Example usage:
+/// ```dart
+/// @StacThemeAnnotation(themeName: 'darkTheme')
+/// ThemeData buildDarkTheme() {
+///   return ThemeData.dark();
+/// }
+/// ```
````
Can we name it `StacThemeConfig`?

divyanshub024 commented on packages/stac_core/lib/annotations/stac_cloud_theme.dart:
````diff
@@ -0,0 +1,19 @@
+///
+/// Example usage:
+/// ```dart
+/// @StacThemeAnnotation(themeName: 'darkTheme')
````
Wrong example here.

divyanshub024 commented on packages/stac/lib/src/services/stac_cache_service.dart:
```diff
@@ -18,13 +16,29 @@ class StacCacheService {
+  /// Gets a cached artifact by its name and type.
   ///
-  /// Returns `null` if the screen is not cached.
-  static Future<StacScreenCache?> getCachedScreen(String screenName) async {
```
I think we should also rename `StacScreenCache` model to `StacCache`

divyanshub024 commented on packages/stac/lib/src/services/stac_cache_service.dart:
```diff
@@ -18,13 +16,29 @@ class StacCacheService {
 
-  /// Gets a cached screen by its name.
+  /// Gets the cache prefix for a given artifact type.
+  static String _getCachePrefix(String artifactType) {
```
Can we make artifactType as enum?

divyanshub024 replied:
I see there is already an enum for `StacArtifactType`, can we use that?

divyanshub024 commented on packages/stac/lib/src/services/stac_cache_service.dart:
```diff
@@ -33,86 +47,60 @@ class StacCacheService {
     } catch (e) {
-      // If there's an error reading from cache, return null
-      // and let the app fetch from network
       return null;
```
Can we add logs wherever we return null? I forgot last time :)

divyanshub024 commented on packages/stac/lib/src/framework/stac_app_theme.dart:
```diff
@@ -0,0 +1,81 @@
+import 'package:stac_core/foundation/theme/stac_theme/stac_theme.dart';
+
+/// Provides helpers to load [StacTheme] definitions for [StacApp].
+class StacAppTheme {
```
Can we move these methods in StacTheme class and use StacTheme.fromCloud

divyanshub024 replied:
Also, we are not using function name `fromCloud` anywhere so we should rename it something else.

Potatomonsta replied:
@divyanshub024 its used in actual app code
```return StacApp(
      title: 'AniWatch',
      theme: StacAppTheme.fromCloud(themeName: "ani_watch_dark"),
      homeBuilder:
          (context) =>
              Stac(routeName: "onboarding_screen"),
      );
```

divyanshub024 replied:
StacAppTheme(name: 'light_theme`)

divyanshub024 commented on packages/stac/lib/src/services/stac_cache_service.dart:
```diff
@@ -121,8 +109,32 @@ class StacCacheService {
+    final cachedArtifact = await getCachedArtifact(artifactName, artifactType);
+    return isCacheValidSync(cachedArtifact, maxAge);
+  }
+
```
Are we using this function anywhere?

divyanshub024 commented on packages/stac/lib/src/framework/stac_app_theme.dart:
```diff
@@ -0,0 +1,81 @@
+      return null;
+    }
+
+    final themePayload = _normalizeThemeJson(rawData['stacJson']);
```
Please change the name of `_normalizeThemeJson` 😆

divyanshub024 commented on packages/stac/lib/src/framework/stac_app_theme.dart:
```diff
@@ -0,0 +1,81 @@
+      return null;
+    }
+    if (payload is Map<String, dynamic> && payload['stacJson'] != null) {
+      return _normalizeThemeJson(payload['stacJson']);
+    }
+    if (payload is Map<String, dynamic>) {
+      return payload;
+    }
+    if (payload is String) {
+      final decoded = jsonDecode(payload);
+      if (decoded is Map<String, dynamic>) {
+        return decoded;
+      }
+    }
```
@Potatomonsta Please fix this one too.

Potatomonsta committed:
Fix: address PR comments

fixes for PR review comments

divyanshub024 commented:
Currently, we need to import all these files. 
```
import 'package:stac_core/annotations/stac_theme_config.dart';
import 'package:stac_core/foundation/colors/stac_brightness.dart';
import 'package:stac_core/foundation/theme/stac_color_scheme/stac_color_scheme.dart';
import 'package:stac_core/foundation/theme/stac_theme/stac_theme.dart';
```
Can we just import the `stac_core/stac_core.dart`

divyanshub024 commented on packages/stac_core/lib/annotations/stac_theme_config.dart:
````diff
@@ -0,0 +1,19 @@
+///   return ThemeData.dark();
+/// }
+/// ```
+class StacThemeConfig {
````
`StacThemeRef`

divyanshub024 commented on packages/stac/lib/src/framework/stac_app_theme.dart:
```diff
@@ -0,0 +1,87 @@
+  /// Fetches a theme from the `/themes` endpoint by [themeName].
+  ///
+  /// Returns `null` if the network call fails or the payload is malformed.
+  static Future<StacTheme?> fromCloud({required String themeName}) async {
+    final response = await StacCloud.fetchTheme(themeName: themeName);
+    if (response == null) {
+      return null;
+    }
+
+    final rawData = response.data;
+    if (rawData is! Map<String, dynamic>) {
```
On error the it keep showing the loader. We should show the default theme.

divyanshub024 replied:
@Potatomonsta On Error we should show log as well the failed to fetch theme from cloud.

divyanshub024 commented:
Add docs and example in movie app

Potatomonsta committed:
feat: Add theming support to documentation and examples

- Introduced a new "Theming" concept in the documentation, detailing how to implement and manage themes in Stac applications.
- Updated the main application example to utilize cloud-based themes instead of local theme definitions.
- Removed the deprecated app_theme.dart file to streamline theme management.
- Enhanced StacApp to support various theme loading methods, including cloud, network, and JSON sources.

divyanshub024 commented:
@Potatomonsta PR checks are failing, please check

Potatomonsta committed:
updated melos version

Potatomonsta committed:
Revert "updated melos version"

This reverts commit 3e98d2c68f071d93a2dbb542e3e8bedc3a7d4006.

divyanshub024 commented on packages/stac/lib/src/framework/stac_app.dart:
```diff
@@ -224,4 +245,80 @@ class StacApp extends StatelessWidget {
+      );
+    }
+    return builder(context, resolved);
+  }
```
@Potatomonsta Convert StacApp to a StatefulWidget and resolve themes in initState()

divyanshub024 commented on packages/stac/lib/src/framework/stac_app.dart:
```diff
@@ -224,4 +245,76 @@ class StacApp extends StatelessWidget {
+  final StacTheme? darkTheme;
+}
+
+class _ThemeFutureLoading extends StatelessWidget {
```
Rename this to `_LoadingWidget`

divyanshub024 commented on packages/stac/lib/src/framework/stac_app.dart:
```diff
@@ -224,4 +245,76 @@ class StacApp extends StatelessWidget {
+  final StacTheme? theme;
+  final StacTheme? darkTheme;
+}
+
```
Can we make this as a record?
`typedef ResolvedStacThemes = ({StacTheme? theme, StacTheme? darkTheme});`

divyanshub024 commented on packages/stac/lib/src/framework/stac_app_theme.dart:
```diff
@@ -0,0 +1,166 @@
+  Future<StacTheme?> resolve() async {
+    switch (_source) {
+      case _ThemeSource.dsl:
+        return Future.value(_dslTheme);
```
Do we need Future.Value? We can return `_dslTheme` directly.

divyanshub024 commented on packages/stac/lib/src/framework/stac_app_theme.dart:
```diff
@@ -0,0 +1,166 @@
+  final _ThemeSource _source;
+  final BuildContext? _context;
+  final StacNetworkRequest? _request;
+  final dynamic _jsonPayload;
```
Can we use Object? for slightly more type safety, since dynamic disables static analysis.

divyanshub024 commented on packages/stac/lib/src/models/stac_artifact_cache.dart:
```diff
@@ -2,15 +2,15 @@ import 'dart:convert';
-class StacScreenCache {
-  /// Creates a [StacScreenCache] instance.
-  const StacScreenCache({
+class StacArtifactCache {
```
Can we rename this to `StacCache`?

divyanshub024 commented on packages/stac/lib/src/models/stac_artifact_cache.dart:
Wrong class name

divyanshub024 commented on packages/stac/lib/src/services/stac_cloud.dart:
```diff
@@ -19,12 +20,41 @@ class StacCloud {
+  /// Tracks artifacts currently being fetched in background to prevent duplicates.
+  static final Map<StacArtifactType, Set<String>> _backgroundFetchInProgress = {
+    StacArtifactType.screen: {},
+    StacArtifactType.theme: {},
+  };
+
+  /// Fetches an artifact from Stac Cloud with intelligent caching.
   ///
```
@Potatomonsta Can we check this?

divyanshub024 commented on packages/stac_core/lib/annotations/stac_theme_ref.dart:
````diff
@@ -0,0 +1,19 @@
+///
+/// Example usage:
+/// ```dart
+/// @StacThemeConfig(themeName: 'darkTheme')
````
Wrong Class name

Potatomonsta committed:
refactor: Update StacApp to StatefulWidget and enhance theme resolution

- Changed StacApp from StatelessWidget to StatefulWidget to manage theme resolution state.
- Introduced a loading state while themes are being resolved.
- Refactored theme resolution logic to improve clarity and efficiency.
- Updated StacArtifactCache to StacCache for consistency across caching models.
- Adjusted related methods and types to reflect the new StacCache naming convention.

Potatomonsta committed:
docs: Refine theming documentation for clarity and accuracy

- Updated the description to clarify the use of DSL and Cloud sources for theme management.
- Removed redundant sections on DSL themes and streamlined the explanation of theme loading methods.
- Adjusted the numbering of theme loading methods for consistency.
- Enhanced best practices and error handling sections for better guidance.

Potatomonsta committed:
switched env var

divyanshub024 reviewed (approved):
LGTM
